### DISCUSSION TITLE
v-form Unexpected Behavior

### DISCUSSION THREAD

**Discussion by connorwinston** [directus/directus#7270] [Archive]
I am creating a custom extension module and am utilizing a v-form component to edit an item in a collection. However, due to the complexities of the permissions with it, I am utilizing custom endpoints for fetching and saving data. I have run into 2 issues when attempting this.

1. Field Input Names Labels do not display when passing field data from the `/fields/{collection-name}` endpoint. However, if you set the collection name it loads the Input Name Labels. I need to remove some fields, so I am trying to just manually pass data.
- Fields prop set with only filtered fields
![image](https://user-images.githubusercontent.com/12714889/128646282-5a2dce10-2961-4a76-9281-3bcbee2e1757.png)
- Collection name set with unwanted fields
![image](https://user-images.githubusercontent.com/12714889/128646314-28f5bde5-fd52-4c68-a95f-8b2fe3aca3bf.png)

2. File upload input successfully uploads file but doesn't update the data once uploaded which results in the interface erroring because it tries to fetch the name of an undefined file.
- Before Uploading File
![image](https://user-images.githubusercontent.com/12714889/128646331-e5a7e63d-0ef0-4393-9081-cc5fbb0739d9.png)
- Console Error
![image](https://user-images.githubusercontent.com/12714889/128646342-9cbc0f78-c9d7-46cc-8d93-46f298a2e1a3.png)
- After Uploading and Error
![image](https://user-images.githubusercontent.com/12714889/128646354-c834b91f-74ea-475e-bdc1-df6edb9620c5.png)

Samples of Code

**Data**
```js
data() {
  return {
    item: null,
    fields: null,
    loading: true,
    edits: null,
  }
},
```
**Load the Fields and Initial Data**
```js
async mounted() {

  const system = inject('system')
  
  let fieldsDenyList = [
    "membership_information",
    "type",
    "status",
    "affiliation",
    "first_name",
    "last_name"
  ]
  
  this.fields = await system.api.get("/fields/members").then((res) => {
    return res.data.data.filter((field) => fieldsDenyList.includes(field.field) === false);
  }).catch((error) => {
    return null
  })
  
  this.item = await system.api.get("/custom/system/members/me").then((res) => {
    return res.data
  }).catch((error) => {
    return null
  })
  this.loading = false
}
```

**Vue Form using Custom Fields**
```vue
<v-form v-model="edits" :fields="fields" :loading="loading" :disabled="loading" :initial-values="item" :primary-key="item.id"  update:model-value="save" />
```

**Comment by rijkvanzanten** [comment]
Heya! 👋🏻 

> Field Input Names Labels do not display when passing field data from the /fields/{collection-name} endpoint. However, if you set the collection name it loads the Input Name Labels. I need to remove some fields, so I am trying to just manually pass data.

This is correct. The app generates the human readable name based on the available information on the field here:

https://github.com/directus/directus/blob/b1f1f6d25b1e458f1d0f55de9de435c845e0f9a7/app/src/stores/fields.ts#L75-L119

It then uses this app-only "pretty name" for rendering things all over the system. In general, I'd recommend pulling the fields from the fields store (`useFieldsStore`) instead of directly from the API, as the store already handles these transformations, and permissions.

> Collection name set with unwanted fields;

If you pass just the collection name to v-form, and no fields, it'll pull the fields from the fields store directly.

> File upload input successfully uploads file but doesn't update the data once uploaded which results in the interface erroring because it tries to fetch the name of an undefined file.

This sounds like a bug to me. Does the same problem occur outside of your custom logic as well?

**Reply by connorwinston** [reply]
Thank you so MUCH! I thought I was loosing my mind trying to figure this out, haha. I will give this a go and then will try some more debugging on the file issue and report back what I have found!

**Reply by connorwinston** [reply]
@rijkvanzanten This is what I have found so far from my digging. The file upload POST, successfully uploads and returns the File ID.
![image](https://user-images.githubusercontent.com/12714889/128790615-bdd8cb80-19db-4f18-bd23-d1d64134aa1d.png)

The file interface is throwing the error...
![image](https://user-images.githubusercontent.com/12714889/128790494-11132d0c-de41-406f-b28f-92e6e414c2ce.png)

It is erroring at 
https://github.com/directus/directus/blob/d94548992ee69d5b0e0b47ab331259140a002d0b/app/src/interfaces/file/file.vue#L20

Then it can't fetch a file with undefined a the ID...
![image](https://user-images.githubusercontent.com/12714889/128790665-4b56816c-f653-424b-bba4-ca7179bb9279.png)

**Reply by rijkvanzanten** [reply]
Thanks for the heads up!

I think this might be a permissions issue 🤔 The fact that `/files` only returns the primary key, and no other format, indicates to me that your user has read access to files, but no read access to any of the other fields. This in turn trips up the v-upload flow, as it expects the file data to be returned at that point

**Reply by connorwinston** [reply]
Ahh, that would be it!!!!! THANK YOU, works like charm now!

**Comment by nickrum** [comment]
Small aside: There is an `@` missing in your provided example template.

```vue
<v-form v-model="edits" ...  @update:model-value="save" />
```

Also, I would suggest either using `v-model` or `:model-value` / `@update:model-value`.

**Reply by connorwinston** [reply]
Thanks for the heads up, I was using the v-model and just forgot that bit was there. Vue is all still little new to me, and I'm getting used to it!
